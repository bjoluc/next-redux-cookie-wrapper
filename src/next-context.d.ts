// Augment next.js AppContext
import { FlushReduxStateToCookies } from ".";
import { NextJSContext } from "next-redux-wrapper";

declare module "next/app" {
  type AppContext = {
    flushReduxStateToCookies: FlushReduxStateToCookies;
    ctx: NextJSContext;
  };
}
