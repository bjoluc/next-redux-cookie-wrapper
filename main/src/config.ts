import {CookieSerializeOptions} from "cookie";
import {Except, SetRequired} from "type-fest";

export type CookieOptions = Except<CookieSerializeOptions, "encode" | "httpOnly">;

export interface SubtreeConfig extends CookieOptions {
	/**
	 * The path of a state subtree that shall be synced with cookies. If, for instance, the state
	 * object is of the form
	 * ```ts
	 * {
	 *   my: {
	 *     config: {...},
	 *   },
	 *   otherState: {...},
	 * }
	 * ```
	 * then some possible `subtree` values would be `my`, `my.config`, and `otherState`.
	 */
	subtree: string;

	/**
	 * Whether or not to ignore a subtree's state in a state update from `getStaticProps()` (defaults
	 * to `true`). If `false`, the state from `getStaticProps()` will be contained in the `HYDRATE`
	 * action without any changes, leaving the merging up to the `HYDRATE` reducer. If `true`, the
	 * received state from `getStaticProps()` will be replaced with the current client state before
	 * the `HYDRATE` action is dispatched. This prevents the client's state from being reset to the
	 * state from `getStaticProps()`.
	 */
	ignoreStateFromStaticProps?: boolean;

	/**
	 * The name that will be assigned to the cookie that holds the subtree's state. Defaults to the
	 * value of the {@link subtree} option.
	 */
	cookieName?: string;

	/**
	 * Whether or not to compress cookie values using lz-string. Defaults to `true`.
	 */
	compress?: boolean;
}

export interface InternalSubtreeConfig
	extends Except<
		SetRequired<SubtreeConfig, "cookieName" | "ignoreStateFromStaticProps" | "compress">,
		keyof CookieOptions
	> {
	cookieOptions: CookieOptions;
}

/**
 * The configuration options for {@link nextReduxCookieMiddleware}. The {@link subtrees} option
 * specifies which subtrees of the state shall be synced with cookies, and how. It takes a list of
 * state subtree paths (e.g. `my.subtree`) or {@link SubtreeConfig} objects. All
 * {@link SubtreeConfig} options (except {@link SubtreeConfig.subtree} and
 * {@link SubtreeConfig.cookieName}) can also be set globally, making them the default for all
 * subtrees.
 *
 * @example
 * ```ts
 * {
 *   secure: true,
 *   subtrees: [
 *     "my.first.subtree",
 *     {subtree: "subtree.two"},
 *     {subtree: "three", secure: false},
 *   ]
 * }
 * ```
 * would set the `secure` option to `true` for the cookies of `my.first.subtree` and `subtree.two`,
 * but `false` for `three`.
 */
export interface NextReduxCookieMiddlewareConfig
	extends Except<SubtreeConfig, "subtree" | "cookieName"> {
	/**
	 * Specifies which subtrees of the state shall be synced with cookies, and how. Takes a list of
	 * subtree paths (e.g. `my.subtree`) and/or {@link SubtreeConfig} objects.
	 */
	subtrees: Array<string | SubtreeConfig>;
}

/**
 * Given a `NextReduxCookieMiddlewareConfig` object, returns the corresponding list of
 * `InternalSubtreeConfig` objects.
 */
export function processMiddlewareConfig(
	config: NextReduxCookieMiddlewareConfig
): InternalSubtreeConfig[] {
	// Set defaults and destructure the config object
	const {subtrees, ...globalSubtreeConfig} = {
		ignoreStateFromStaticProps: true,
		compress: true,
		path: "/",
		sameSite: true,
		...config,
	};

	return (
		subtrees
			// Turn strings into objects, set a default for the cookieName option, apply the global
			// default config, and extract cookie options into separate objects
			.map((current) => {
				if (typeof current === "string") {
					current = {subtree: current};
				}

				const {ignoreStateFromStaticProps, compress, subtree, cookieName, ...cookieOptions} = {
					...globalSubtreeConfig,
					cookieName: current.subtree,
					...current,
				};
				return {
					ignoreStateFromStaticProps,
					compress,
					subtree,
					cookieName,
					cookieOptions,
				};
			})
	);
}
