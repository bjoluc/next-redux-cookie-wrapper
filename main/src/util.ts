import {GetServerSidePropsContext, NextPageContext} from "next";
import {Context} from "next-redux-wrapper";
import {AppContext} from "next/app";
import {SetRequired} from "type-fest";

export const isClient = () => typeof window !== "undefined";

/**
 * Returns whether the provided context is a `GetServerSidePropsContext` or a `PageContext` with
 * `req` and `res` set.
 */
export const isSsrContext = (
	context: Context
): context is GetServerSidePropsContext | SetRequired<NextPageContext, "req" | "res"> =>
	(context as GetServerSidePropsContext).req !== undefined &&
	(context as GetServerSidePropsContext).res !== undefined;

/**
 * Returns whether the provided context is an `AppContext`.
 */
export const isAppContext = (context: Context): context is AppContext =>
	(context as AppContext).ctx !== undefined;
