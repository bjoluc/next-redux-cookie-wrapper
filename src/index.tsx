import {IncomingMessage, ServerResponse} from "http";

import ServerCookies from "cookies";
import {NextComponentType, NextPageContext} from "next";
import withRedux, {
  MakeStore,
  MakeStoreOptions,
  Config as NextReduxWrapperConfig,
  WrappedAppProps,
} from "next-redux-wrapper";
import {AppContext} from "next/app";
import * as React from "react";
import {Action, AnyAction, Store, createStore} from "redux";
import {
  PersistConfig,
  Persistor,
  createPersistoid,
  getStoredState,
  persistReducer,
  persistStore,
} from "redux-persist";
// @ts-expect-error No type definitions and we do not want to create a global definition in this package
import {CookieStorage, NodeCookiesWrapper} from "redux-persist-cookie-storage";
import {Except} from "type-fest";

export type CustomPersistConfig<S> = Except<PersistConfig<S>, "storage" | "key"> &
  Partial<Pick<PersistConfig<S>, "key">>;

export type Config = NextReduxWrapperConfig & {
  persistConfig?: CustomPersistConfig<any>;
  cookieConfig?: any;
};

export type FlushReduxStateToCookies = () => Promise<void>;

const defaultPersistConfig = {
  key: "root",
  blacklist: [] as string[],
};

const defaultConfig: Config = {
  persistConfig: defaultPersistConfig,
  cookieConfig: {},
};

export const withReduxCookiePersist = (makeStore: MakeStore, config?: Config) => {
  config = {
    ...defaultConfig,
    ...config,
  };
  const {persistConfig, cookieConfig, ...reduxWrapperConfig} = config;

  const sharedPersistConfig = {
    ...defaultPersistConfig,
    ...persistConfig,
  };

  const sharedCookieConfig = {
    setCookieOptions: {},
    ...cookieConfig,
  };

  const debug = reduxWrapperConfig.debug ?? false;

  const extractStateFromCookies = async (req: IncomingMessage, res: ServerResponse) => {
    // @ts-ignore https://github.com/ScottHamper/Cookies/pull/83
    const cookies = new NodeCookiesWrapper(new ServerCookies(req, res));

    const persistConfig = {
      ...sharedPersistConfig,
      storage: new CookieStorage(cookies, sharedCookieConfig),
    };

    let state: any | undefined;
    try {
      state = await getStoredState(persistConfig);
    } catch (error) {
      /* istanbul ignore if */
      if (debug) {
        console.log(
          "getStoredState() failed (this happens when the index storage item is not set):\n",
          error
        );
      }
    }

    // Removing the state's _persist key for the server-side (non-persisted) redux store
    if (state && typeof state._persist !== "undefined") {
      const {_persist, ...cleanedState} = state;
      state = cleanedState;
    }

    return state;
  };

  // Used internally by flushReduxStateToCookies()
  const createPersistor = async (store: Store): Promise<Persistor> =>
    new Promise((resolve) => {
      const persistor = persistStore(store, {}, () => {
        resolve(persistor);
      });
    });

  async function flushReduxStateToCookies(this: NextPageContext) {
    /* istanbul ignore if */
    if (!(this.req && this.res)) {
      if (debug) {
        console.log(
          "flushReduxStateToCookies only works on the server " +
            "in your app's getInitialProps() method. Ignoring the call."
        );
      }

      return;
    }

    /* istanbul ignore if */
    if (debug) {
      console.log("Flushing the store's current state to cookies via flushReduxStateToCookies...");
    }

    // @ts-ignore https://github.com/ScottHamper/Cookies/pull/83
    const cookies = new NodeCookiesWrapper(new ServerCookies(this.req, this.res));
    const persistConfig = {
      ...sharedPersistConfig,
      blacklist: [...sharedPersistConfig.blacklist, "_persist"],
      storage: new CookieStorage(cookies, {
        ...sharedCookieConfig,
        setCookieOptions: {
          ...sharedCookieConfig.setCookieOptions,
          httpOnly: false, // Allow modifications on the client side
        },
      }),
      stateReconciler(_inboundState: any, originalState: any) {
        // Ignore state from cookies, only use the store's current state
        return originalState;
      },
    };

    // Using a dummy reducer here as we do not dispatch actions to this store
    const reducer = persistReducer(persistConfig, (state, _action) => state);
    const store = createStore(reducer, this.store.getState());
    const persistor = await createPersistor(store);

    // Set cookies
    await persistor.flush();

    /* istanbul ignore if */
    if (debug) {
      console.log("State flushed to cookies:", store.getState());
    }
  }

  const wrappedMakeStore = (initialState: any, options: MakeStoreOptions) => {
    if (options.req && options.res) {
      initialState = (options.req as any).__initialReduxCookieState;
    }

    const store = makeStore(initialState, options);

    if (!options.isServer) {
      // Let's persist the store!
      const persistConfig = {
        ...sharedPersistConfig,
        storage: new CookieStorage(require("cookies-js"), sharedCookieConfig),
      };

      // Note: We do not create a persistor here because we need no rehydration.
      // See https://github.com/rt2zz/redux-persist/issues/457#issuecomment-362490893 for the idea
      const persistoid = createPersistoid(persistConfig);
      store.subscribe(() => {
        persistoid.update(store.getState());
      });
    }

    return store;
  };

  return (App: NextComponentType | any) => {
    // Wrap the provided App with next-redux-wrapper
    const WrappedApp = withRedux(wrappedMakeStore, reduxWrapperConfig)(App);

    return class CookiePersistedWrappedApp extends React.Component<WrappedAppProps> {
      public static displayName = `withReduxCookiePersist(${WrappedApp.displayName})`;

      public static getInitialProps = async (appCtx: AppContext) => {
        if (appCtx.ctx.req && appCtx.ctx.res) {
          /* istanbul ignore if */
          if (debug) {
            console.log("0. Extracting state from cookies (if any)");
          }

          const state = await extractStateFromCookies(appCtx.ctx.req, appCtx.ctx.res);
          /* istanbul ignore if */
          if (debug) {
            console.log("0. Got state", state);
          }

          // Illegally attaching the extracted state to the request object so we can access it later
          // on in our makeStore wrapper
          (appCtx.ctx.req as any).__initialReduxCookieState = state;
        }

        appCtx.ctx.flushReduxStateToCookies = flushReduxStateToCookies;

        return WrappedApp.getInitialProps(appCtx);
      };

      public render() {
        return <WrappedApp {...this.props} />;
      }
    };
  };
};

// Augment Next.js NextPageContext
declare module "next/dist/next-server/lib/utils" {
  export interface NextPageContext<S = any, A extends Action = AnyAction> {
    /**
     * Provided by next-redux-cookie-wrapper: If the code is executed on the server and `ctx.req`
     * and `ctx.res` are set, this method will add a cookies header with the redux store's current
     * state to `ctx.res`.
     *
     * Note: This will only take effect if you redirect the client afterwards. Otherwise, do not
     * call this function â€“ the client will use the up-to-date serialized state contained in the
     * HTML response anyway and overwrite the cookies with it.
     */
    flushReduxStateToCookies: FlushReduxStateToCookies;
  }
}
