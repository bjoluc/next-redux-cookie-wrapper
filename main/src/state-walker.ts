import {produce} from "immer";
import get from "lodash/get";
import set from "lodash/set";
import {JsonObject} from "type-fest";

import {InternalSubtreeConfig} from "./config";

/**
 * Given a list of subtree configurations, two state objects `stateA` and `stateB`, and a walker
 * function, invokes the walker function for each `InternalSubtreeConfig` object. The walker
 * function receives a subtree config and the respective state subtrees from `stateA` and `stateB`.
 * If a subtree does not exist in a state object, `undefined` is passed instead.
 *
 * @param subtrees A list of `InternalSubtreeConfig` objects for the walker function
 * @param stateA First state object
 * @param stateB Optional second state object. If omitted the walker function will always receive
 * `undefined` as its last parameter.
 * @param walker A function that will be invoked for each `InternalSubtreeConfig` object from
 * `subtrees` with the `InternalSubtreeConfig` object and the respective state subtrees from
 * `stateA` and `stateB`.
 *
 * @returns A copy of `stateA` where those subtrees for which the walker function has returned a
 * value have been replaced by that value.
 */
export const walkState = <State extends JsonObject>(
	subtrees: InternalSubtreeConfig[],
	walker: (
		subtreeConfig: InternalSubtreeConfig,
		subtreeA?: unknown,
		subtreeB?: unknown
	) => unknown | undefined,
	stateA: State,
	stateB?: State
) =>
	// @ts-ignore https://github.com/immerjs/immer/issues/839
	produce(stateA, (draftState) => {
		for (const subtreeConfig of subtrees) {
			const {subtree} = subtreeConfig;
			const result = walker(subtreeConfig, get(stateA, subtree), get(stateB, subtree));
			if (typeof result !== "undefined") {
				set(draftState, subtree, result);
			}
		}
	});
