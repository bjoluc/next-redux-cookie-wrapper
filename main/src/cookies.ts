import {compressToEncodedURIComponent, decompressFromEncodedURIComponent} from "lz-string";
import {GetServerSidePropsContext, NextPageContext} from "next";
import {parseCookies, setCookie} from "nookies";
import {JsonValue, SetRequired} from "type-fest";

import {InternalSubtreeConfig} from "./config";
import {isClient} from "./util";

export type Cookies = Record<string, JsonValue>;

export type CookieContext = SetRequired<
	Partial<GetServerSidePropsContext | NextPageContext>,
	"req" | "res"
>;

export type CookieConfig = Pick<InternalSubtreeConfig, "cookieName" | "compress" | "cookieOptions">;

/**
 * An isomorphic class to set and get (compressed) state cookies.
 */
export class StateCookies {
	protected _config = new Map<string, CookieConfig>();

	private readonly _context?: CookieContext;
	private _cookies?: Cookies;

	/**
	 * @param context The Next.js context for the request (if on the server)
	 */
	constructor(context?: CookieContext) {
		this._context = context;
	}

	private static _encodeState(state: any) {
		return encodeURIComponent(JSON.stringify(state));
	}

	private static _encodeStateCompressed(state: any) {
		return compressToEncodedURIComponent(JSON.stringify(state));
	}

	private static _decodeState(state: string, compressed: boolean) {
		return JSON.parse(
			(compressed ? decompressFromEncodedURIComponent : decodeURIComponent)(state)!
		);
	}

	/**
	 * Set the configuration (@see CookieConfig) for each cookie
	 */
	public setConfigurations(configurations: CookieConfig[]) {
		for (const config of configurations) {
			this._config.set(config.cookieName, config);
		}
	}

	public getAll() {
		// Parse cookies if they have not been parsed, always re-parse cookies on the client
		if (typeof this._cookies === "undefined" || isClient()) {
			this._cookies = {};
			for (const [name, value] of Object.entries(
				parseCookies(this._context, {decode: (value: string) => value})
			)) {
				const config = this._config.get(name);
				if (config) {
					this._cookies[name] = StateCookies._decodeState(value, config.compress);
				}
			}
		}

		return this._cookies;
	}

	public set(name: string, state: any) {
		const {cookieOptions, compress} = this._config.get(name)!;
		setCookie(this._context, name, state, {
			...cookieOptions,
			encode: compress ? StateCookies._encodeStateCompressed : StateCookies._encodeState,
			httpOnly: false,
		});
	}
}
