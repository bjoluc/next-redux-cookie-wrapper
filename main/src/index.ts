import isEqual from "fast-deep-equal";
import {Context, HYDRATE, MakeStore} from "next-redux-wrapper";
import {AnyAction, Middleware, Store} from "redux";

import {NextReduxCookieMiddlewareConfig, processMiddlewareConfig} from "./config";
import {StateCookies} from "./cookies";
import {walkState} from "./state-walker";
import {isAppContext, isClient, isSsrContext} from "./util";

export {NextReduxCookieMiddlewareConfig, SubtreeConfig} from "./config";

/**
 * An action type that is dispatched internally on the server to pass the Next.js context to the
 * {@link nextReduxCookieMiddleware}.
 */
export const SERVE_COOKIES = "next-redux-cookie-wrapper/SERVE_COOKIES";

/**
 * A tiny wrapper for the `makeStore()` function. It sends the received Next.js context to
 * {@link nextReduxCookieMiddleware} on the server via a {@link SERVE_COOKIES} action.
 *
 * Use it like this:
 * ```ts
 * export const wrapper = createWrapper<AppStore>(wrapMakeStore(makeStore));
 * ```
 *
 * @param makeStore The `makeStore()` function for `next-redux-wrapper`
 * @returns A wrapped version of the `makeStore()` function
 */
export const wrapMakeStore =
	<S extends Store>(makeStore: MakeStore<S>) =>
	(context: Context) => {
		const store = makeStore(context);

		if (isClient()) {
			// Dispatch an empty HYDRATE action in case the current page doesn't have getServerSideProps,
			// getStaticProps, or getInitialProps and there are state cookies available. The middleware
			// will then add any state subtrees from cookies to the HYDRATE action's payload.
			store.dispatch({type: HYDRATE, payload: {}});
		} else {
			if (isAppContext(context)) {
				context = context.ctx;
			}

			if (isSsrContext(context)) {
				// Create StateCookies object and smuggle it to the middleware via the `SERVE_COOKIES` action
				store.dispatch({
					type: SERVE_COOKIES,
					payload: new StateCookies(context),
				});
			}
		}

		return store;
	};

/**
 * A Redux middleware that syncs user-defined subtrees of the Redux state with cookies â€“ on the
 * server and on the client. One cookie is used per state subtree and the serialized state is, by
 * default, compressed using [lz-string](https://github.com/pieroxy/lz-string). The subtree paths,
 * cookie names, cookie options, and compression are configured via a
 * {@link NextReduxCookieMiddlewareConfig} object.
 */
export const nextReduxCookieMiddleware: (config: NextReduxCookieMiddlewareConfig) => Middleware =
	(config) => (store) => {
		const subtrees = processMiddlewareConfig(config);

		let cookies: StateCookies;
		if (isClient()) {
			cookies = new StateCookies();
			cookies.setConfigurations(subtrees);
		}
		// On the server, we have to intercept the `SERVE_COOKIES` action to get the `cookies` object
		// (we cannot directly set a property on the store, sadly, since the middleware does not have
		// direct access to the store object).

		return (next) => (action: AnyAction) => {
			switch (action.type) {
				case SERVE_COOKIES: {
					// Handle the SERVE_COOKIES action (server-only):
					cookies = action.payload;
					cookies.setConfigurations(subtrees);
					// Console.log("Cookies received by middleware");

					// We have access to the client's cookies now. Now we need to hydrate the store with their
					// state.
					const allCookies = cookies.getAll();

					// Console.log("Triggering initial HYDRATE");
					store.dispatch({
						type: HYDRATE,
						payload: walkState(subtrees, (subtree) => allCookies[subtree.cookieName], {}),
					});

					// We're done handling the action here (without calling `next()`).
					return;
				}

				case HYDRATE:
					if (isClient()) {
						// Intercept next-redux-wrapper's HYDRATE action on the client

						const allCookies = cookies.getAll();
						action.payload = walkState(
							subtrees,
							({ignoreStateFromStaticProps, cookieName}) => {
								if (ignoreStateFromStaticProps) {
									// `action.payload` holds the incoming server state. We overwrite that state with
									// the state from the cookies: If the incoming state is from getServerSideProps,
									// the cookies have also been updated to that state. If the incoming state is from
									// getStaticProps, the cookies have remained unchanged and hence the server's
									// state is ignored.

									return allCookies[cookieName];
								}
							},
							action.payload
						);
					}
				// Fall through to default handling of the `HYDRATE` action

				default: {
					const oldState = store.getState();
					const result = next(action);
					const newState = store.getState();

					// If cookies are available (which is not the case during `getStaticProps()`), write the
					// new state into cookies wherever necessary
					if (cookies) {
						walkState(
							subtrees,
							({cookieName}, oldSubtreeState, newSubtreeState) => {
								if (!isEqual(oldSubtreeState, newSubtreeState)) {
									cookies.set(cookieName, newSubtreeState);
								}
							},
							oldState,
							newState
						);
					}

					return result;
				}
			}
		};
	};
