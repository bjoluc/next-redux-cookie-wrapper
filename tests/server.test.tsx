/**
 * @jest-environment node
 **/

import httpMocks from "node-mocks-http";
// @ts-ignore No type definitions and we do not want to create a global definition in this package
import { NodeCookiesWrapper } from "redux-persist-cookie-storage";

import { withReduxCookiePersist } from "../src/";
import { FlushStateStorePage, PlainApp, StoreApp, makeStore, verifyComponent } from "./util";

let appCtx: any;

let setCookie: jest.Mock;

beforeEach(() => {
  NodeCookiesWrapper.prototype.get = jest.fn();

  setCookie = jest.fn();
  NodeCookiesWrapper.prototype.set = setCookie;

  const req = httpMocks.createRequest({ url: "/" });
  const res = httpMocks.createResponse({ req });

  appCtx = {
    ctx: {
      pathname: "/",
      req,
      res,
      query: {},
    },
  };
});

describe("withReduxCookiePersist", () => {
  it("should integrate the store using next-redux-wrapper", async () => {
    let WrappedApp = withReduxCookiePersist(makeStore)(PlainApp);
    await verifyComponent(WrappedApp, appCtx, "init");

    WrappedApp = withReduxCookiePersist(makeStore)(StoreApp);
    await verifyComponent(WrappedApp, appCtx, "foo");
  });

  describe("without incoming cookies", () => {
    it("should pass initialState = undefined to makeStore() at its first call", async () => {
      const makeStoreSpy = jest.fn(makeStore);
      const WrappedApp = withReduxCookiePersist(makeStoreSpy)(PlainApp);

      await verifyComponent(WrappedApp, appCtx, "init");

      expect(makeStoreSpy).toHaveBeenCalledTimes(2);
      expect(makeStoreSpy.mock.calls[0][0]).toBeUndefined();
    });
  });

  describe("with incoming cookies", () => {
    beforeEach(() => {
      const cookies: { [key: string]: string } = {
        "persist:root": JSON.stringify({
          reduxStatus: JSON.stringify("fromCookie"),
          _persist: JSON.stringify({ version: -1, rehydrated: true }),
        }),
        reduxPersistIndex: '["persist:root"]',
      };
      NodeCookiesWrapper.prototype.get = jest.fn((key) => cookies[key]);
    });

    it("should use the cookies' state and pass it to makeStore() at its first call", async () => {
      const makeStoreSpy = jest.fn(makeStore);
      const WrappedApp = withReduxCookiePersist(makeStoreSpy)(PlainApp);

      await verifyComponent(WrappedApp, appCtx, "fromCookie");

      expect(makeStoreSpy).toHaveBeenCalledTimes(2);
      expect(makeStoreSpy.mock.calls[0][0]).toEqual({ reduxStatus: "fromCookie" });
    });
  });

  describe("with malformed incoming cookies", () => {
    beforeEach(() => {
      const cookies: { [key: string]: string } = {
        "persist:root": JSON.stringify({
          reduxStatus: "notAJsonString",
        }),
        reduxPersistIndex: '["persist:root"]',
      };
      NodeCookiesWrapper.prototype.get = jest.fn((key) => cookies[key]);
    });

    it("should use the default state (pass undefined to makeStore)", async () => {
      const makeStoreSpy = jest.fn(makeStore);
      const WrappedApp = withReduxCookiePersist(makeStoreSpy)(PlainApp);

      await verifyComponent(WrappedApp, appCtx, "init");

      expect(makeStoreSpy).toHaveBeenCalledTimes(2);
      expect(makeStoreSpy.mock.calls[0][0]).toBeUndefined();
    });
  });

  describe("flushReduxStateToCookies() called within getInitialProps()", () => {
    describe.each([true, false])("with incoming cookies: %p", (withStateCookies) => {
      it("should set cookies using NodeCookieWrapper", async () => {
        if (withStateCookies) {
          const cookies: { [key: string]: string } = {
            "persist:root": JSON.stringify({
              reduxStatus: JSON.stringify("fromCookie"),
              _persist: JSON.stringify({ version: -1, rehydrated: true }),
            }),
            reduxPersistIndex: '["persist:root"]',
          };
          NodeCookiesWrapper.prototype.get = jest.fn((key) => cookies[key]);
        }

        const WrappedApp = withReduxCookiePersist(makeStore)(FlushStateStorePage);
        await WrappedApp.getInitialProps(appCtx);

        if (!withStateCookies) {
          // The reduxPersistIndex is not set again if it exists (which is when withStateCookies is true)
          expect(setCookie).toHaveBeenCalledWith("reduxPersistIndex", '["persist:root"]', {
            httpOnly: false,
          });
        }
        expect(setCookie).toHaveBeenCalledWith(
          "persist:root",
          JSON.stringify({
            reduxStatus: JSON.stringify("foo"),
          }),
          { httpOnly: false }
        );
      });
    });
  });
});
