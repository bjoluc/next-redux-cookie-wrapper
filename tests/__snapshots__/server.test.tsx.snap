// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withReduxCookiePersist should integrate the store using next-redux-wrapper 1`] = `
<div>
  {"isServer":true,"custom":"custom"}
  {"reduxStatus":"init"}
</div>
`;

exports[`withReduxCookiePersist should integrate the store using next-redux-wrapper 2`] = `
<div>
  {"isServer":true,"custom":"custom"}
  {"reduxStatus":"foo"}
</div>
`;

exports[`withReduxCookiePersist with incoming cookies should use the cookies' state and pass it to makeStore() at its first call 1`] = `
<div>
  {"isServer":true,"custom":"custom"}
  {"reduxStatus":"fromCookie"}
</div>
`;

exports[`withReduxCookiePersist with malformed incoming cookies should use the default state (pass undefined to makeStore) 1`] = `
<div>
  {"isServer":true,"custom":"custom"}
  {"reduxStatus":"init"}
</div>
`;

exports[`withReduxCookiePersist without incoming cookies should pass initialState = undefined to makeStore() at its first call 1`] = `
<div>
  {"isServer":true,"custom":"custom"}
  {"reduxStatus":"init"}
</div>
`;
